Veilid Chunking

rasswanth-s opened the pull request:
Re-opening @itstauq  veilid chunking PR.

## Description
Veilid can send at most 32 kb of data in one network call. However the requests-responses from PySyft can be larger than than. This PR contains utilities to send and receive larger data by chunking the data into packets of max 32 kb.

### Items to cover in later cuts:

- Timeout the buffers to prevent memory leaks.
- Write unit and integration tests
- Reliability improvements and error recovery (resending failed chunks or handling network interruptions gracefully).
- What happens in case the servers are behind a load balancer? The VeilidStreamer request can land in any instance but the buffer will not be shared among the instances as everything is in-memory. 
    - Does this even needs to be handled?
    - Idea 1: Use sticky connections in the load balancer.
    - Idea 2: Use a pull-based approach for sending the chunks. Sender sends start stream request -> Receiver makes subsequent requests to get all the chunks from the sender.

This PR closes https://github.com/OpenMined/Heartbeat/issues/1076

## Affected Dependencies
None

## How has this been tested?
- [x] Using tests present in `notebooks/Testing/Veilid/Large-Message-Testing.ipynb`

## Checklist
- [x] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [x] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [x] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [ ] My changes are covered by tests

itstauq committed:
add VeilidStreamer for sending large messages in chunks

itstauq committed:
Refactor everything

itstauq committed:
fix bugs in veilid streamer and add notebooks for testing

itstauq committed:
Make VeilidStreamer compatible with app_call by adding callback to receive_stream
Make notebooks compatible with new VeilidStreamer and refactor
Lots of other refactoring and cleanup

itstauq committed:
Add instructions to use VeilidStreamer in the docstring

itstauq committed:
Send chunks parallelly, add batching and retries, improve test notebooks

itstauq committed:
fix linting issues

itstauq committed:
fix linting issues try 2

itstauq committed:
fix linting issues try 3

itstauq committed:
add architecture diagram for VeilidStreamer

itstauq committed:
Integrate VeilidStreamer with veilid_core
Make VeilidStreamer a singleton
Add notebooks to test sending large messages using /app_call endpoint

itstauq committed:
Add code to benchmark Veilid's performance with different message sizes

itstauq committed:
Add a single notebook with instructions to send large veilid messages

itstauq committed:
replace batching and retries with semaphores; update test notebook

itstauq committed:
- Use random UUID to identify a stream call instead of message hash
- Make code more robust and add support for sending messages above 16 MB
- Refactor and make code cleaner

itstauq committed:
Merge branch 'rasswanth/veilid-prototype' into tauquir/veilid-chunking

rasswanth-s committed:
Merge branch 'dev' into tauquir/veilid-chunking

madhavajay committed:
Merge branch 'dev' into tauquir/veilid-chunking

itstauq committed:
Add ability to send responses greater than 32 KB in size

itstauq committed:
Merge remote-tracking branch 'origin' into tauquir/veilid-chunking

itstauq committed:
Merge branch 'tauquir/veilid-chunking' of github.com:OpenMined/PySyft into tauquir/veilid-chunking

itstauq committed:
Make VeilidStreamer compatible with vld_key and update notebooks

itstauq committed:
Add retries to VeilidStreamer; remove unnecessary notebooks

itstauq committed:
Add endpoint for testing handle_app_call and add more tests to notebook

rasswanth-s commented on packages/grid/veilid/server/veilid_core.py:
```diff
@@ -151,7 +153,11 @@ async def app_call(vld_key: str, message: bytes) -> bytes:
 
-            result = await router.app_call(route, message)
+            result = (
+                await router.app_call(route, message)
```
I think , we might also need to handle the case, when the request could be less than 32 KB, but the response might be a large value.

rasswanth-s commented on packages/grid/veilid/server/veilid_streamer.py:
```diff
@@ -0,0 +1,393 @@
+
+    def _calculate_chunks_count(self, message_size: int) -> int:
+        max_chunk_size = self.chunk_size - self.stream_chunk_header_struct.size
+        total_no_of_chunks = message_size // max_chunk_size + 1
```
MAYBE: I think this formula could be 
as 
```python3
total_no_of_chunks = math.ceil( message_size / max_chunk_size)
```
as the former formula would fail if the message size is perfectly divisible.

example:

message size = 10
chunk size = 2
total chunks = 5 (the previous formula would return 6)

rasswanth-s commented on packages/grid/veilid/server/veilid_streamer.py:
```diff
@@ -0,0 +1,393 @@
+        else:
+            logger.error(f"[Bad Message] Message with unknown prefix: {prefix}")
+
+    @retry(veilid.VeilidAPIError, tries=4, delay=1, backoff=2)
```
❤️

itstauq committed:
- Abstract away app_call and app_call_reply entirely
- Improve performance for single chunk streams

itstauq committed:
add changes for resolving PR comments

itstauq committed:
Merge branch 'dev' into tauquir/veilid-chunking

itstauq committed:
Update docstring for STREAM_SINGLE request

rasswanth-s commented on packages/grid/veilid/server/veilid_streamer.py:
```diff
@@ -0,0 +1,441 @@
+CallId = bytes
+
+
+class RequestType(bytes, ReprEnum):
```
Q: What is the use of ReprEnum?

itstauq replied:
This will create a sort of `BytesEnum` class, similar to the inbuilt `enum.IntEnum` and `enum.StrEnum`. Helps in validation and allows directly accessing members instead of needing to use `enum_member.value` in all places (otherwise we would need to do `RequestType.STREAM_SINGLE.value` and `RequestType.STREAM_CHUNK.value` instead of `RequestType.STREAM_SINGLE` and `RequestType.STREAM_CHUNK`.

To further improve readability, I'll add a separate `BytesEnum` class and inherit the `RequestType` class from that instead.

rasswanth-s replied:
Ahh ok makes sense

rasswanth-s commented on packages/grid/veilid/server/veilid_streamer.py:
```diff
@@ -0,0 +1,441 @@
+
+class ResponseType(bytes, ReprEnum):
+    OK = b"@VS@OK"
+    ERROR = b"@VS@ER"
```
MAYBE:  we could keep the bytes value: 
```python3
b"@VS@ERR"
```

itstauq replied:
I wanted to keep the length same for both the `ResponseType`s

itstauq committed:
Add BytesEnum to utils.py and use it in veilid_streamer.py

rasswanth-s commented on packages/grid/veilid/server/veilid_streamer.py:
```diff
@@ -0,0 +1,441 @@
+        return cls._instance
+
+    def __init__(self) -> None:
+        self.chunk_size = MAX_MESSAGE_SIZE
```
MAYBE: we could rename this variable to 
```python3
self.max_message_size 
```
because in other places we  use chunk_size for formula calculation.

itstauq replied:
Sure, that'll make more sense! How about renaming it to `self.max_single_veilid_message_size`?

itstauq replied:
Because I think the former might imply a global maximum message size limit...

rasswanth-s commented on packages/grid/veilid/server/veilid_streamer.py:
```diff
@@ -0,0 +1,441 @@
+            call_id,
+            chunk_number,
+        )
+        max_actual_message_size = self.chunk_size - self.stream_chunk_header_struct.size
```
MAYBE: we could re-name to a common variable name (for max_actual_message_size)
we do the same formula calculation at function `_calculate_chunks_count`
but the variable name is `max_chunk size`

rasswanth-s replied:
Also, we could have the max chunk size calculation as a separate method or present as an instance variable in the class for uniformity?

itstauq replied:
Adding a `self.max_data_in_each_stream_chunk_request` instance variable

rasswanth-s replied:
Keep variable name more descriptive is great.but we could try shortening like 
maybe
1. max_stream_single_msg_size
2. max_stream_chunk_msg_size

rasswanth-s commented on packages/grid/veilid/server/veilid_streamer.py:
```diff
@@ -0,0 +1,477 @@
+
+    @retry(veilid.VeilidAPIError, tries=4, delay=1, backoff=2)
+    async def _send_request(
+        self, router: veilid.RoutingContext, dht_key: str, request_data: bytes
```
MAYBE: Could we re-name in all the places from dht_key to vld_key with the new notation

rasswanth-s replied:
for the streamer part

itstauq committed:
Add changes for resolving PR comments; fix logging

rasswanth-s commented on packages/grid/veilid/server/veilid_streamer.py:
```diff
@@ -0,0 +1,480 @@
+        # https://docs.python.org/3/library/struct.html#format-characters
+        BYTE_ORDER = "!"  # big-endian is recommended for networks as per IETF RFC 1700
+        REQUEST_TYPE_PREFIX = f"{RequestType.SIZE}s"
+        CALL_ID = "16s"
```
MAYBE: On reading the code, since veilid also has a call_id attribute for each app call and we use it in functions like _send_response. To make it easy for new users navigating the code.

We could re-name  our `call_id` to `stream_id` ?

itstauq replied:
This was super needed 🙌

rasswanth-s commented on packages/grid/veilid/server/veilid_streamer.py:
```diff
@@ -0,0 +1,480 @@
+        chunk_header, chunk = message[:chunk_header_len], message[chunk_header_len:]
+        _, call_id, chunk_number = self.stream_chunk_header_struct.unpack(chunk_header)
+        buffer = self.buffers[call_id]
+        buffer.receive_chunk(chunk_number, chunk)
```
MAYBE: we could re-name the method as `buffer.add_chunk` ?

rasswanth-s commented on packages/grid/veilid/server/veilid_streamer.py:
```diff
@@ -0,0 +1,480 @@
+        self.chunks: list[bytes | None]
+        self.message: asyncio.Future[bytes] = asyncio.Future()
+        self.holds_reply: bool = holds_reply
+        # TODO add mechanism to delete/timeout old buffers
```
MAYBE: we could have another variable like
self.chunks_received : int

which stores the number of chunks received.

This could be incremented in the `receive_chunk` method when we add each chunk.

During handling stream end request before `re-assembling` , we could use this number
to 


```python3
if len(self.chunks) != self.chunks_received
       raise Exception("did not receive all the chunks")
```

until we have Error handling and Re-transmission integrated in Future.

itstauq replied:
It does makes sense to add error-handling for empty chunks 💯 
However, since receiving chunks is a highly parallel operation, adding an incrementing integer will introduce the need to add locks, adding synchronization overhead and potentially slowing down the whole process. Instead, we can implement a check for empty chunks while handling the stream end request before re-assembling. How does that sound?

rasswanth-s replied:
Good catch, we could do it at the stream end 💯

rasswanth-s commented on packages/grid/veilid/server/veilid_streamer.py:
```diff
@@ -0,0 +1,480 @@
+            cls._instance.buffers = {}
+        return cls._instance
+
+    def __init__(self) -> None:
```
MAYBE: The __init__ method has a lot of initialization, we could something like this

```python3
def __init__(self) -> None:
        self._init_structs()
        self._init_message_sizes()
        self._init_semaphores()

def _init_structs(self) -> None:
    BYTE_ORDER = "!"  
    REQUEST_TYPE_PREFIX = f"{RequestType.SIZE}s"
    CALL_ID = "16s"
    MESSAGE_HASH = "32s"
    TOTAL_CHUNKS_COUNT = "Q"
    CURRENT_CHUNK_NUMBER = "Q"

    self.stream_start_struct = Struct(
        BYTE_ORDER
        + REQUEST_TYPE_PREFIX
        + CALL_ID
        + MESSAGE_HASH
        + TOTAL_CHUNKS_COUNT
    )
    self.stream_chunk_header_struct = Struct(
        BYTE_ORDER + REQUEST_TYPE_PREFIX + CALL_ID + CURRENT_CHUNK_NUMBER
    )
    self.stream_end_struct = Struct(BYTE_ORDER + REQUEST_TYPE_PREFIX + CALL_ID)
    self.stream_single_struct = Struct(BYTE_ORDER + REQUEST_TYPE_PREFIX + CALL_ID)

def _init_message_sizes(self) -> None:
    self.max_single_veilid_message_size = MAX_SINGLE_VEILID_MESSAGE_SIZE
    self.max_data_in_each_stream_chunk_request = (
        self.max_single_veilid_message_size - self.stream_chunk_header_struct.size
    )
    self.max_data_in_each_stream_single_request = (
        self.max_single_veilid_message_size - self.stream_single_struct.size
    )

def _init_semaphores(self) -> None:
    self._send_request_semaphore = asyncio.Semaphore(MAX_STREAMER_CONCURRENCY)
    self._send_response_semaphore = asyncio.Semaphore(MAX_STREAMER_CONCURRENCY)
```

rasswanth-s commented on packages/grid/veilid/server/veilid_streamer.py:
```diff
@@ -0,0 +1,480 @@
+            buffer_for_holding_reply = Buffer(holds_reply=True)
+            self.buffers[call_id] = buffer_for_holding_reply
+
+        if message_size <= self.max_data_in_each_stream_single_request:
```
We could do a split of the functions to keep it more readable like

```python3
if message_size <= self.max_single_chunk_size:
    await self._stream_single_chunk_request(router, vld_key, message, stream_id)
else:
    await self._stream_multi_chunk_request(router, vld_key, message, stream_id)

if is_request_stream:
    await self._wait_for_reply(stream_id)


```

rasswanth-s commented on packages/grid/veilid/server/veilid_streamer.py:
```diff
@@ -0,0 +1,480 @@
+            if response != ResponseType.OK:
+                raise Exception("Unexpected response from server")
+
+    async def _send_response(
```
MAYBE: we could also add retry logic to responses?

itstauq replied:
Retrying a failed `app_call_reply` makes Veilid throw some operation id mismatch errors. We need to retry the original request in that case.

rasswanth-s replied:
Ahh ok 👍 .We could file a bug with veilid team on the same later?

itstauq replied:
Yep 💯

itstauq committed:
Add suggestions from PR comments
- Change Call ID to Stream ID throughout VeilidStreamer
- Clean up VeilidStreamer.__init__ method by moving the initialization of structs, message sizes, and semaphores to separate methods
- Clean up the VeilidStreamer.stream method by moving the logic to wait for reply and sending single and multi chunk requests to separate methods
- Add a new method to cleanup the buffer after the stream is complete
- Other minor changes

itstauq committed:
Ensure that all the chunks are received before reassembling the message

itstauq committed:
Merge branch 'dev' into tauquir/veilid-chunking

rasswanth-s committed:
removed compression proxy route
fix peer client caching with node uid + route

rasswanth-s committed:
Merge branch 'dev' into tauquir/veilid-chunking

rasswanth-s commented:
Stellar Work @itstauq  ⭐ 
It's great to Chunking working with great reliability
![test](https://github.com/OpenMined/PySyft/assets/43314053/b2e33150-6838-4a91-aa14-2bb77db9ea88)

rasswanth-s committed:
skip veilid tests temporarily
